access-common/src/main/java/edu/unc/lib/dl/ui/service/LorisContentService.java

```diff
@@ -167,23 +176,140 @@ public void streamJP2(String simplepid, String region, String size, String rotat
+
+        String title = getTitle(rootObj);
+
+        Manifest manifest = new Manifest(URIUtil.join(manifestBase, "manifest"), title);
+
+        String abstractText = rootObj.getAbstractText();
+        String label = rootObj.getLabel();
+        List<String> creators = rootObj.getCreator();
+        List<String> subjects = rootObj.getSubject();
+        List<String> language = rootObj.getLanguage();
+
+        if (abstractText != null) {
+            manifest.addDescription(abstractText);
+        }
+
+        if (label != null) {
+            manifest.addLabel(label);
+        }
+
+        manifest.addLogo(new ImageContent(URIUtil.join(basePath, "static", "images", "unc-icon.png")));
+
+        setMetadataField(manifest, "Creators", creators);
+        setMetadataField(manifest, "Subjects", subjects);
+        setMetadataField(manifest, "Languages", language);
+        manifest.addMetadata("", "<a href=\"" +
+                URIUtil.join(basePath, "record", rootObj.getId()) + "\">View full record</a>");
+        String attribution = "University of North Carolina Libraries, Digital Collections Repository";
+        String collection = rootObj.getParentCollectionName();
+        if (collection != null) {
+            attribution += " - Part of " + collection;
+        }
+        manifest.addMetadata("Attribution", attribution);
+
+        Sequence seq = createSequence(manifestBase, briefObjs);
+
+        return iiifMapper.writeValueAsString(manifest.addSequence(seq));
+    }
+
+    public String getSequence(HttpServletRequest request, List<BriefObjectMetadata> briefObjs)
+            throws JsonProcessingException {
+        String path = getRecordPath(request);
+        return iiifMapper.writeValueAsString(createSequence(path, briefObjs));
+    }
+
+    public String getCanvas(HttpServletRequest request, BriefObjectMetadata briefObj)
+            throws JsonProcessingException {
+        String path = getRecordPath(request);
+        return iiifMapper.writeValueAsString(createCanvas(path, briefObj));
+    }
+
+    private Sequence createSequence(String seqPath, List<BriefObjectMetadata> briefObjs) {
+        Sequence seq = new Sequence(URIUtil.join(seqPath, "sequence", "normal"));
+
+        Set<String> uuidList = new HashSet<>();
+        for (BriefObjectMetadata briefObj : briefObjs) {
+            String datastreamUuid = jp2Pid(briefObj);
+
+            if (!datastreamUuid.equals("")) {
+                // Don't add rootObj twice
+                if (uuidList.contains(datastreamUuid)) {
+                    continue;
+                }
+
+                Canvas canvas = createCanvas(seqPath, briefObj);
+                seq.addCanvas(canvas);
+            }
+
+            uuidList.add(datastreamUuid);
+        }
+
+        return seq;
+    }
+
+    private Canvas createCanvas(String path, BriefObjectMetadata briefObj) {
+        String title = getTitle(briefObj);
+        String uuid = jp2Pid(briefObj);
```

Comment:
if there's no JP2 for the object, I think this method is going to produce an invalid canvas path (uuid will either be blank or null). This only impacts `getCanvas`, since the usage in `createSequence` is already verifying that there is a jp2. For `getCanvas`, i'm not sure what the best approach would be.  Returning the canvas without adding any of the jp2/thumbnail stuff? Or throwing an exception?